File: src/db/seed.ts
import { db } from "#/db"
import { env } from "#/env"
import { auth } from "#/lib/auth"

export const seed = async () => {
  console.log("Seeding database...")

  const user = await db.query.user.findFirst()
  if (!user) {
    console.log("Creating admin user...")
    const res = await auth.api.createUser({
      body: {
        name: "Admin",
        email: env.ADMIN_EMAIL,
        password: "password",
        role: "admin",
        data: { username: "admin" },
      },
    })
    console.log(
      "Admin user created with id",
      res.user.id,
      "and email",
      res.user.email,
      "and default password is 'password'. Please change the password after logging in.",
    )
  }

  console.log("Database seeded")
}

File: internal/db/seed.go
package db

import (
	"context"
	"database/sql"
	"log/slog"
	"time"
	"uuid"

	"github.com/zareix/dockstack/internal/auth"
	"github.com/zareix/dockstack/internal/db/store"
)

func Seed(ctx context.Context, sqlDB *sql.DB, adminEmail string) error {
	q := store.New(sqlDB)
	count, err := q.CountUsers(ctx)
	if err != nil {
		return err
	}
	if count > 0 {
		return nil
	}
	hash, err := auth.HashPassword("password")
	if err != nil {
		return err
	}
	userID := uuid.New().String()
	username := "admin"
	now := time.Now().UnixMilli()
	err = q.CreateUser(ctx, store.CreateUserParams{
		ID:            userID,
		Name:          "Admin",
		Email:         adminEmail,
		EmailVerified: true,
		Username:      &username,
		Role:          "admin",
		CreatedAt:     now,
		UpdatedAt:     now,
	})
	if err != nil {
		return err
	}
	err = q.UpsertCredential(ctx, store.UpsertCredentialParams{
		UserID:       userID,
		PasswordHash: hash,
		UpdatedAt:    now,
	})
	if err != nil {
		return err
	}
	slog.Warn("seeded admin user with default password \"password\" — change it after first login", "email", adminEmail)
	return nil
}
